dht_rt_seeder.go

```diff
@@ -0,0 +1,119 @@
+package dht
+
+import (
+	"context"
+	"time"
+
+	"github.com/ipfs/go-todocounter"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// SeederDialTimeout is the grace period for one dial attempt
+var SeederDialTimeout = 5 * time.Second
+
+// TotalSeederDialTimeout is the total grace period for a group of dial attempts
+var TotalSeederDialTimeout = 30 * time.Second
+
+// SeederConcurrentDials is the number of peers we will dial simultaneously
+var SeederConcurrentDials = 50
+
+func (dht *IpfsDHT) seedRoutingTable(candidates, fallbacks []peer.ID) error {
+	peersFromSet := func(m map[peer.ID]struct{}) []peer.ID {
+		var peers []peer.ID
+		for p := range m {
+			peers = append(peers, p)
+		}
+
+		return peers
+	}
+
+	// candidates we have still not tried dialling to
+	remainingCandidates := make(map[peer.ID]struct{})
+	// fallbacks we have still not tried dialling to
+	remainingFallbacks := make(map[peer.ID]struct{})
+
+	for i := range candidates {
+		remainingCandidates[candidates[i]] = struct{}{}
+	}
+
+	for i := range fallbacks {
+		remainingFallbacks[fallbacks[i]] = struct{}{}
+	}
+
+	for {
+
+		todial := dht.seedsProposer.Propose(dht.routingTable, peersFromSet(remainingCandidates), peersFromSet(remainingFallbacks))
+		if len(todial) == 0 {
+			logger.Infof("dht rt seeder: finished as proposer returned nil, RT size is now %d", dht.routingTable.Size())
+			return nil
+		} else {
+			// start dialing
+			type result struct {
+				p   peer.ID
+				err error
+			}
+
+			// attempts to dial to a given peer to verify it's available
+			dialFn := func(ctx context.Context, p peer.ID, res chan<- result) {
+				childCtx, cancel := context.WithTimeout(ctx, SeederDialTimeout)
+				defer cancel()
+				_, err := dht.host.Network().DialPeer(childCtx, p)
+				select {
+				case <-ctx.Done(): // caller has already hung up & gone away
+				case res <- result{p, err}:
+				}
+			}
+
+			resCh := make(chan result) // dial results.
+			ctx, cancel := context.WithTimeout(dht.ctx, TotalSeederDialTimeout)
+			defer cancel()
+
+			// start dialing
+			semaphore := make(chan struct{}, SeederConcurrentDials)
+			go func(peers []peer.ID) {
+				for _, p := range peers {
+					semaphore <- struct{}{}
+					go func(p peer.ID, res chan<- result) {
+						dialFn(ctx, p, res)
+						<-semaphore
+					}(p, resCh)
+				}
+			}(todial)
+
+			// number of results we expect on the result channel
+			todo := todocounter.NewSyncCounter()
+			todo.Increment(uint32(len(todial)))
+
+		LOOP:
+			for {
+				select {
+				case res := <-resCh:
+					todo.Decrement(1)
+					if res.err != nil {
+						logger.Infof("dht rt seeder: discarded proposed peer due to dial error; peer ID: %s, err: %s", res.p, res.err)
+					} else {
+						if _, err := dht.routingTable.Update(res.p); err != nil {
```

Comment:
Need to document  the fact that till #283 is fixed, this could result into two calls to `rt.Update(P)` as dialling & connecting to `P` will also cause a`rt.Update(p)` call if the dht notification service has already started.

However, this does not cause any unwanted side effects for now.